File: python/src/ouster/sdk/client/scan_source.py
from typing import Iterator, Union, Optional, List
from typing_extensions import Protocol
from ouster.sdk._bindings.client import SensorInfo, LidarScan, FieldType


class ScanSource(Protocol):
    """Represents only data stream from one stream."""

    @property
    def metadata(self) -> SensorInfo:
        """A 'SensorInfo' object associated with the scan streams."""
        ...

    @property
    def is_live(self) -> bool:
        """True if data obtained from the RUNNING sensor or as a stream from the socket

        Returns:
            True if data obtained from the RUNNING sensor or as a stream from the socket
            False if data is read from a stored media. Restarting an ``iter()`` means that
                    the data can be read again.
        """
        ...

    @property
    def is_seekable(self) -> bool:
        """True for non-live sources, This property can be True regardless of scan source being indexed or not.
        """
        ...

    @property
    def is_indexed(self) -> bool:
        """True for IndexedPcap and OSF scan sources, this property tells users whether the underlying source
        allows for random access of scans, see __getitem__.
        """
        ...

    @property
    def field_types(self) -> List[FieldType]:
        """Field types are present in the LidarScan objects on read from iterator"""
        ...

    @property
    def fields(self) -> List[str]:
        """Fields are present in the LidarScan objects on read from iterator"""
        ...

    @property
    def scans_num(self) -> Optional[int]:
        """Number of scans available, in case of a live sensor or non-indexable scan source
         this method returns None"""
        ...

    def __len__(self) -> int:
        """Number of scans available, in case of a live sensor or non-indexable scan source this method
         throws a TypeError"""
        ...

    def __iter__(self) -> Iterator[LidarScan]:
        ...

    def _seek(self, key: int) -> None:
        """seek/jump to a specific item within the list of LidarScan objects that this particular scan
        source has access to"""
        ...

    # NOTE: based on the underlying implemention the return type is
    # Optional[LidarScan] since MultiScanSource returns collate scans by default.
    # This can be solved by provide a method that gives access to uncollated scans
    def __getitem__(self, key: Union[int, slice]
                    ) -> Union[Optional[LidarScan], 'ScanSource']:
        """Indexed access and slices support"""
        ...

    def close(self) -> None:
        """Manually release any underlying resource."""
        ...

    def __del__(self) -> None:
        """Automatic release of any underlying resource."""
        ...

    def _slice_iter(self, key: slice) -> Iterator[Optional[LidarScan]]:
        ...

    def slice(self, key: slice) -> 'ScanSource':
        """Constructs a ScanSource matching the specificed slice"""
        ...

File: ouster_client/src/scan_source.cpp
#include "ouster/scan_source.h"

namespace ouster {
namespace core {
std::vector<std::shared_ptr<LidarScan>> ScanSource::operator[](
    int start) const {
    auto iter = begin();
    if (start < 0) {
        auto len = end() - begin();
        iter += len + start;
    } else {
        iter += start;
    }
    if (iter == end()) {
        throw std::out_of_range("Indexed past the end of the scan source.");
    }
    return {*iter};
}

size_t ScanSource::size() const { return end() - begin(); }

ScanIterator ScanSource::end() const { return ScanIterator(this); }

bool ScanSource::is_live() const { return false; }

bool ScanSource::is_indexed() const { return false; }

const std::vector<std::pair<uint64_t, uint64_t>>& ScanSource::full_index()
    const {
    throw std::runtime_error(
        "'full_index' not supported on unindexed scan sources.");
}

const std::vector<std::vector<std::pair<uint64_t, uint64_t>>>&
ScanSource::individual_index() const {
    throw std::runtime_error(
        "'individual_index' not supported on unindexed scan sources.");
}

const std::vector<size_t>& ScanSource::scans_num() const {
    throw std::runtime_error(
        "'scans_num' not supported on unindexed scan sources.");
}

}  // namespace core
}  // namespace ouster
